icaruscode/CRT/CRTTPCTruthEffMCKalman_module.cc

```diff
@@ -0,0 +1,1721 @@
+		yi2 += (hit_points[i].Y() - yavg)*(hit_points[i].Y() - yavg);
+		zi2 += (hit_points[i].Z() - zavg)*(hit_points[i].Z() - zavg);
+		xiy += (hit_points[i].X() - xavg)*(hit_points[i].Y() - yavg);
+		xiz += (hit_points[i].X() - xavg)*(hit_points[i].Z() - zavg);
+		yiz += (hit_points[i].Y() - yavg)*(hit_points[i].Z() - zavg);
+*/
+
+	}//end loop over points in data
+	xi2 = xi2/hit_points.size(); yi2 = yi2/hit_points.size(); zi2 = zi2/hit_points.size();
+	xiy = xiy/hit_points.size(); xiz = xiz/hit_points.size(); yiz = yiz/hit_points.size();
+
+	Eigen::Matrix3d covmat; 
+	covmat << xi2, xiy, xiz, xiy, yi2, yiz, xiz, yiz, zi2;
+
+	Eigen::SelfAdjointEigenSolver<Eigen::Matrix3d> eigenMat(covmat);
+
+	std::cout << "covariant matrix:\n" << covmat << std::endl;
+
+	if(eigenMat.info() == Eigen::ComputationInfo::Success){
+
+	      reco::PrincipalComponents::EigenValues recobEigenVals = eigenMat.eigenvalues().cast<float>();
+	      reco::PrincipalComponents::EigenVectors recobEigenVecs = eigenMat.eigenvectors().cast<float>();
+
+	      std::cout << "Eigenvalues are: " << recobEigenVals(0) << ", " << recobEigenVals(1) << ", " << recobEigenVals(2) << std::endl;
+
+	      float maxeval = FLT_MIN; int maxvalpos = -1;
+
+	      for(int tempindex=0; tempindex<3; tempindex++){
+
+			if(maxeval<(float)recobEigenVals(tempindex)) {
+
+				maxeval = (float)recobEigenVals(tempindex);
+				maxvalpos = tempindex;
+
+			}//end if(maxeval<(float)recobEigenVals)
+
+	      }//end for(int tempindex=0; tempindex<3; tempindex++)
+
+	      bestfit_dir.SetXYZ(recobEigenVecs(0,maxvalpos), recobEigenVecs(1,maxvalpos), recobEigenVecs(2,maxvalpos));
+
+	}//end if(eigenMat.info() == Eigen::ComputationInfo::Success)
+
+	}//end if(!forcedbreak)
+	std::cout << "End endpoint_PCA_ana function............\n-----------------------------------------\n\n\n";
+
+}//end definition of void endpoint_PCA_ana(recob::Track trk, int startorend, TVector3 &bestfit_dir, TVector3 &bestfit_pos)
+
+void icarus::CRTTPCTruthEffMCKalman::pointChecker(recob::Track trk, bool &startisinTPC, bool &endisinTPC, int &startflag, int &endflag){
+
+	geo::Point_t startpoint = trk.Start();
+	geo::Point_t endpoint = trk.End();
+
+	float xmin = 61.94;    float xmax = 358.49;
+	float ymin = -181.86;  float ymax = 134.96;
+	float zmin = -894.951; float zmax = 894.951;
+
+	bool x_start_check, x_end_check, y_start_check, y_end_check, z_start_check, z_end_check;
+
+	x_start_check = std::abs(startpoint.X())>=xmin && std::abs(startpoint.X())<=xmax;
+	x_end_check = std::abs(endpoint.X())>=xmin && std::abs(endpoint.X())<=xmax;
+	y_start_check = startpoint.Y()>=ymin && startpoint.Y()<=ymax;
+	y_end_check = endpoint.Y()>=ymin && endpoint.Y()<=ymax;
+	z_start_check = startpoint.Z()>=zmin && startpoint.Z()<=zmax;
+	z_end_check = endpoint.Z()>=zmin && endpoint.Z()<=zmax;
+
+	startisinTPC = x_start_check && y_start_check && z_start_check;
+	endisinTPC = x_end_check && y_end_check && z_end_check;
+
```

Comment:
It would be good to note somewhere what the different values of startflag and endflag mean.